### PR TITLE
WIP (DO NOT MERGE YET): sendgrid v3

### PR DESCRIPTION
**Pull request by tomohara**

### CODE REVIEW

File: espwrap/adaptors/sendgrid_v3.py [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+            if self.track_clicks:
+                tracking_settings.click_tracking = ClickTracking(
+                    self.track_clicks,
+                    self.track_clicks
+                )
+            message.tracking_settings = tracking_settings
+            
+            #Personalizations (DO THIS LAST)
+            for subgrp in to_send:
+                p=0
```
Comment (jimbattin) [inline espwrap/adaptors/sendgrid_v3.py]: you can have python do this automatically with `enumerate`
eg: `for p, recip in enumerate(subgrp)` -- I would use a better variable name than p, which technically isn't PEP8 compliant.

File: espwrap/adaptors/sendgrid_v3.py [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+                response = self.client.send(message)
+                print("status code")
+                print(response.status_code)
+                print("body")
+                print(response.body)
+                print("headers")
+                print(response.headers)
+            except exceptions.BadRequestsError as e:
+                print(e.body)
+                exit()
```
Comment (jimbattin) [inline espwrap/adaptors/sendgrid_v3.py]: I'd replace exit with raise. You generally don't want libraries to ever throw exit codes. It's up to the person using your library to handle this error situation.

File: espwrap/adaptors/sendgrid_v3.py [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+from python_http_client import exceptions
+
+
+from espwrap.base import MassEmail, batch, MIMETYPE_HTML, MIMETYPE_TEXT
+
+
+if sys.version_info < (3,):
+    range = xrange
+
+import os
```
[the remark this answers is not in the document]
Comment (tomohara) [reply espwrap/adaptors/sendgrid_v3.py]: thanks for this. i did a cleanup of imports ordering and groupings, and removed of the un-needed/un-used range=xrange block.

File: espwrap/adaptors/sendgrid_v3.py [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+        if email_no_alias in seen:
+            to_send.append([recip])
+        else:
+            seen.add(email_no_alias)
+            to_send[0].append(recip)
+
+    return to_send
+
+
+class SendGridMassEmail(MassEmail):
```
[the remark this answers is not in the document]
Comment (tomohara) [reply espwrap/adaptors/sendgrid_v3.py]: the X-Message-ID is returned in the header of the response from mail send.
not yet being captured and stored, but doing so is in the plans.

File: espwrap/adaptors/sendgrid_v3.py [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.INFO)
+handler = logging.StreamHandler(sys.stdout)
+handler.setLevel(logging.INFO)
+formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
+handler.setFormatter(formatter)
+logger.addHandler(handler)
+
+
+def breakdown_recipients(grp):
```
[the remark this answers is not in the document]
Comment (tomohara) [reply espwrap/adaptors/sendgrid_v3.py]: it is the exact same. now it's DRY in a common module.

File: espwrap/adaptors/sendgrid_v3.py [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+                    message.subject = Subject(self.subject)
+
+                    for key, val in recip['merge_vars'].items():
+                        new_key = '{1}{0}{2}'.format(key, *self.delimiters)
+                        message.substitution = Substitution(new_key, val, p=p)
+                    p+=1
+
+            try:
+                response = self.client.send(message)
+                print("status code")
```
[the remark this answers is not in the document]
Comment (tomohara) [reply espwrap/adaptors/sendgrid_v3.py]: created a basic verbose flag for the class (default False)

File: espwrap/adaptors/sendgrid_v3.py (hunk 1 above)
Comment (tomohara) [reply espwrap/adaptors/sendgrid_v3.py]: nice. i set it this way, getting "p" for free.

File: espwrap/adaptors/sendgrid_v3.py (hunk 2 above)
Comment (tomohara) [reply espwrap/adaptors/sendgrid_v3.py]: agreed and replaced. i borrowed this block from sendgrid repo use cases.
now uses raise. will evolve from here ... not sure what other exceptions need to be addressed, but the plan is to at least manage sendgrid API limitations and log when the payload is too big, too many personalizations, etc.

**Pull request closed without merging**